Add Phoenix and TruLens third-party scorer integrations

debu-sinha opened the pull request:
### Related Issues/PRs

Partial fix for #19062 (TruLens portion)

**Note:** This PR has been split per reviewer feedback. Phoenix scorers are now in #19473.

### What changes are proposed in this pull request?

This PR adds integration for TruLens evaluation framework as MLflow GenAI scorers.

**TruLens scorers (`mlflow.genai.scorers.trulens`):**
- `Groundedness`: Evaluates groundedness in context
- `ContextRelevance`: Assesses context relevance to query
- `AnswerRelevance`: Evaluates answer relevance to query
- `Coherence`: Evaluates coherence and logical flow

**Implementation details:**
- Lazy loading to avoid import overhead when scorers not used
- Returns `CategoricalRating.YES/NO` values with scores in metadata
- `get_scorer()` API for dynamic metric selection
- Configurable model providers (OpenAI, Databricks, LiteLLM)
- Trace input support for extracting context from retrieval spans
- Clear error messages when optional dependencies are missing

---

## Usage Examples with Real Output

### Direct Scorer Call

```python
from mlflow.genai.scorers.trulens import Groundedness

scorer = Groundedness(model="openai:/gpt-4o-mini")
feedback = scorer(
    outputs="The Eiffel Tower is 330 meters tall.",
    expectations={"context": "The Eiffel Tower is 330 metres tall and located in Paris."},
)
print(f"Value: {feedback.value}")
print(f"Metadata: {feedback.metadata}")
```

**Output:**
```
Value: yes
Metadata: {'mlflow.scorer.framework': 'trulens', 'score': 1.0, 'threshold': 0.5}
```

### With mlflow.genai.evaluate()

```python
import mlflow
from mlflow.genai.scorers.trulens import Groundedness, AnswerRelevance

eval_data = [
    {
        "inputs": {"query": "What is the capital of France?"},
        "outputs": "Paris is the capital of France.",
        "expectations": {"context": "France is in Western Europe. Its capital is Paris."},
    },
    {
        "inputs": {"query": "What is the capital of France?"},
        "outputs": "London is the capital of France.",  # Wrong - should fail
        "expectations": {"context": "France is in Western Europe. Its capital is Paris."},
    },
]

scorers = [
    Groundedness(model="openai:/gpt-4o-mini"),
    AnswerRelevance(model="openai:/gpt-4o-mini"),
]

results = mlflow.genai.evaluate(data=eval_data, scorers=scorers)
print(results.metrics)
```

**Output:**
```
{'Groundedness/mean': 0.5, 'AnswerRelevance/mean': 0.5}
```

### All 4 TruLens Scorers

```python
from mlflow.genai.scorers.trulens import Groundedness, ContextRelevance, AnswerRelevance, Coherence

all_scorers = [
    Groundedness(model="openai:/gpt-4o-mini"),
    ContextRelevance(model="openai:/gpt-4o-mini"),
    AnswerRelevance(model="openai:/gpt-4o-mini"),
    Coherence(model="openai:/gpt-4o-mini"),
]
results = mlflow.genai.evaluate(data=eval_data, scorers=all_scorers)
```

**Output:**
```
AnswerRelevance/mean: 0.5000
Coherence/mean: 0.5000
ContextRelevance/mean: 1.0000
Groundedness/mean: 0.5000
```

### get_scorer() API

```python
from mlflow.genai.scorers.trulens import get_scorer

groundedness = get_scorer("Groundedness", model="openai:/gpt-4o-mini")
context_relevance = get_scorer("ContextRelevance", model="databricks")
answer_relevance = get_scorer("AnswerRelevance", model="databricks:/my-endpoint")
coherence = get_scorer("Coherence", model="openai:/gpt-4o-mini")
```

---

### How is this PR tested?

- [x] New unit tests in `tests/genai/scorers/trulens/test_trulens.py`
- [x] Real API integration tests with OpenAI (all tests passed)

Test coverage includes:
- All 4 scorers with positive and negative test cases
- Edge cases (empty strings, long text, special characters)
- `get_scorer()` API for all metrics
- `mlflow.genai.evaluate()` batch evaluation
- `CategoricalRating.YES/NO` value validation
- `AssessmentSource.LLM_JUDGE` type validation
- Error handling for missing required fields

### Does this PR require documentation update?

- [x] Yes. I've updated:
  - [x] API references with comprehensive docstrings

### Release Notes

#### Is this a user-facing change?

- [x] Yes

Added TruLens third-party scorer integration for MLflow GenAI evaluation. 4 new scorers available:

**TruLens scorers:** Groundedness, ContextRelevance, AnswerRelevance, Coherence

Install dependencies:
```bash
pip install trulens trulens-providers-openai
```

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/evaluation`: MLflow model evaluation features

#### How should the PR be classified in the release notes?

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [ ] Yes
- [x] No (this PR will be included in the next minor release)

debu-sinha commented:
@AveshCSingh Would you be able to take a look at this PR when you have a chance? It adds third-party scorer integrations for Phoenix (Arize) and TruLens evaluation frameworks, enabling seamless use of established LLM evaluation tools within MLflow's GenAI evaluation pipeline. Happy to address any feedback.

debu-sinha committed:
Add Phoenix and TruLens third-party scorer integrations

Implements support for Phoenix (Arize) and TruLens evaluation frameworks
as MLflow GenAI scorers, enabling seamless integration of established
LLM evaluation tools within MLflow's evaluation pipeline.

Phoenix scorers (mlflow/genai/scorers/phoenix/):
- PhoenixHallucinationScorer: Detects hallucinations in model outputs
- PhoenixRelevanceScorer: Evaluates response relevance to queries
- PhoenixToxicityScorer: Assesses content toxicity
- PhoenixQAScorer: Evaluates QA correctness
- PhoenixSummarizationScorer: Assesses summarization quality

TruLens scorers (mlflow/genai/scorers/trulens/):
- TruLensGroundednessScorer: Evaluates groundedness in context
- TruLensContextRelevanceScorer: Assesses context relevance
- TruLensAnswerRelevanceScorer: Evaluates answer relevance
- TruLensCoherenceScorer: Evaluates logical flow of outputs

Features:
- Lazy loading to avoid import overhead when not used
- Configurable model providers (OpenAI, LiteLLM for TruLens)
- Consistent scorer interface returning Feedback objects
- Helpful error messages when optional dependencies missing

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

joelrobin18 commented on mlflow/genai/scorers/__init__.py:
```diff
@@ -40,15 +40,34 @@
 }
 
+# Third-party scorer integrations (Phoenix, TruLens)
+_THIRDPARTY_IMPORTS = {
```
Instead of third party, it would be better to split them into two different folders
One for Phoenix and other for Trulens

That would be better instead of using third party.

debu-sinha commented:
Hi @joelrobin18,

Thanks for the feedback on the folder structure. I've restructured the code as suggested:

- Created separate `mlflow/genai/scorers/phoenix/` directory
- Created separate `mlflow/genai/scorers/trulens/` directory  
- Each integration has its own `__init__.py` with proper exports
- Updated the lazy loading in the main `scorers/__init__.py` accordingly

Would appreciate your re-review when you get a chance. Also cc @AveshCSingh for visibility.

debu-sinha commented:
Note: I also contributed a fix to the TruLens project (https://github.com/truera/trulens/pull/2308) to add `additionalProperties: false` to Pydantic schemas for Databricks structured output compatibility. This was merged today and ensures TruLens works correctly with Databricks endpoints - relevant context for this integration.

debu-sinha committed:
Fix Phoenix score handling and add validation for both integrations

Key changes:
1. Phoenix scorers: Remove incorrect score inversion for Hallucination and
   Toxicity evaluators. Phoenix natively returns scores aligned with MLflow
   convention (1.0 = good, 0.0 = bad).

2. Both Phoenix and TruLens scorers: Replace silent score clamping with
   validation that logs warnings when scores are outside expected 0-1 range.
   This helps detect potential version incompatibilities.

3. Update tests to reflect correct Phoenix score semantics:
   - factual = 1.0 (not inverted from 0)
   - non-toxic = 1.0 (not inverted from 0)

4. Update docstrings and comments to accurately describe score semantics.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

B-Step62 commented on mlflow/genai/scorers/phoenix/phoenix.py:
```diff
@@ -0,0 +1,455 @@
+        from phoenix.evals import OpenAIModel
+
+        return OpenAIModel(model=self.model_name or "gpt-4o-mini")
+
```
Phoenix support several different models not only OpenAI: https://arize-phoenix.readthedocs.io/en/arize-phoenix-v4.10.1/api/evals.models.html

Can we support them? LiteLLM model will be fallback for models that are not natively supported by them e.g., Gemini.

B-Step62 commented on mlflow/genai/scorers/phoenix/phoenix.py:
```diff
@@ -0,0 +1,455 @@
+
+                logging.getLogger(__name__).warning(
+                    f"Phoenix returned score {normalized_score} outside expected 0-1 range. "
+                    "This may indicate a version incompatibility. Clamping to valid range."
```
Do we need to crop the score to 0-1?

joelrobin18 replied:
Do we have any example on when the score wasn't between 0-1?

B-Step62 commented on mlflow/genai/scorers/phoenix/phoenix.py:
```diff
@@ -0,0 +1,455 @@
+        inputs: dict[str, Any] | None = None,
+        outputs: Any | None = None,
+        context: str | None = None,
+        **kwargs,
```
A scorer should only have `inputs`, `outputs`, `expectations`, and `trace` as arguments. Other things like context should be derived from one of these.

B-Step62 commented on mlflow/genai/scorers/phoenix/phoenix.py:
```diff
@@ -0,0 +1,455 @@
+        # Already aligned with MLflow convention (higher = better)
+        score, rationale = self._parse_result(result, positive_label="factual")
+
+        return Feedback(name=self.name, value=score, rationale=rationale)
```
Let's set assessment source as well.

smoorjani commented on mlflow/genai/scorers/phoenix/phoenix.py:
small update - we called it `get_scorer()` now

smoorjani reviewed (changes requested):
left some nits in addition to Yuki's comments - I think the general ask here is to follow the existing pattern from DeepEval/RAGAS where we have a generalized implementation for any scorer (`get_scorer`) and the classes themselves (e.g., `FaithfulnessScorer`) are super lightweight wrappers.

debu-sinha committed:
Refactor Phoenix and TruLens scorers to follow get_scorer() pattern

- Align with existing DeepEval/RAGAS patterns for consistency
- Add get_scorer() API for both Phoenix and TruLens integrations
- Create lightweight wrapper classes (Hallucination, Relevance, etc.)
- Add multi-model support: Databricks managed judge, serving endpoints, LiteLLM
- Use standard scorer arguments: inputs, outputs, expectations, trace
- Add AssessmentSource for proper attribution
- Use CategoricalRating.YES/NO for binary feedback values
- Add models.py, registry.py, utils.py for each integration
- Update tests to match new API patterns

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented on mlflow/genai/scorers/phoenix/phoenix.py:
```diff
@@ -0,0 +1,455 @@
+
+        return OpenAIModel(model=self.model_name or "gpt-4o-mini")
+
+    def _parse_result(
```
Agreed - I've refactored this entirely. The  method now directly constructs and returns a  object with proper . No separate helper method needed since each scorer's evaluation logic is self-contained.

debu-sinha committed:
Fix Phoenix and TruLens test mocking

- Use function-level patching instead of module-level mocking
- Import classes before applying mocks to ensure module is loaded
- Fix line length issues in test files

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Fix Toxicity scorer field mapping for output-only calls

Phoenix ToxicityEvaluator expects text in the 'input' field, but MLflow
scorers typically pass text via 'outputs' parameter. When Toxicity is
called with only outputs (no inputs), the text should be mapped to
Phoenix's 'input' field.

This fix ensures Toxicity()(outputs='text') works correctly by mapping
the output to Phoenix's expected 'input' field when no input is provided.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Fix Databricks managed judge adapters for Phoenix and TruLens

The call_chat_completions function from databricks.rag_eval requires
the @context.eval_context decorator which sets up internal state. When
called directly from Phoenix/TruLens model adapters without this context,
it fails with 'cannot access local variable' errors.

Switch to using _invoke_databricks_serving_endpoint with the
databricks-meta-llama-3-3-70b-instruct foundation model endpoint,
which works in both notebook and external environments.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Add _verbose and _rate_limiter attributes required by Phoenix

Phoenix evaluators use a set_verbosity context manager that expects
model adapters to have both _verbose and _rate_limiter._verbose attributes.
Without these, the context manager fails with UnboundLocalError.

Added:
- _NoOpRateLimiter stub class with _verbose attribute
- _verbose = False to both Databricks model adapter classes
- _rate_limiter = _NoOpRateLimiter() to both classes

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Fix Phoenix and TruLens Databricks adapters for framework compatibility

Phoenix:
- Add _verbose attribute required by set_verbosity context manager
- Add _rate_limiter with _NoOpRateLimiter stub for set_verbosity

TruLens:
- Create dynamic provider class that inherits from LLMProvider
- Implement _create_chat_completion method required by TruLens
- TruLens feedback methods (groundedness, relevance, etc.) are inherited
  from LLMProvider base class

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Fix Phoenix MultimodalPrompt and TruLens Pydantic initialization

Phoenix fix:
- Handle MultimodalPrompt objects by converting to string before passing
  to Databricks endpoint (Phoenix evaluators pass MultimodalPrompt, not str)

TruLens fix:
- Properly initialize Pydantic base class with super().__init__(model_engine=...)
- Use class-level attributes for endpoint config to avoid Pydantic field issues

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Fix TruLens provider endpoint initialization

Add Endpoint object to TruLens Databricks provider initialization.
TruLens LLMProvider requires an endpoint to be set for feedback methods
like groundedness_measure_with_cot_reasons to work correctly.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
Renewed tests on Databricks:
<img width="1130" height="768" alt="Screenshot 2025-12-17 at 2 27 06 PM" src="https://github.com/user-attachments/assets/4a46fc76-9ec0-4ca8-bb27-5ad483ac464f" />
<img width="1037" height="721" alt="Screenshot 2025-12-17 at 2 27 14 PM" src="https://github.com/user-attachments/assets/a95ec7ce-b4db-4cb4-915c-16e1f8ba2920" />
<img width="918" height="566" alt="Screenshot 2025-12-17 at 2 27 22 PM" src="https://github.com/user-attachments/assets/6f26cf0f-6276-4d45-85d8-24a966656172" />
<img width="1298" height="472" alt="Screenshot 2025-12-17 at 2 27 28 PM" src="https://github.com/user-attachments/assets/e5c7d0bf-92d0-416b-a4ed-ca80b527fc2b" />
<img width="677" height="541" alt="Screenshot 2025-12-17 at 2 27 33 PM" src="https://github.com/user-attachments/assets/c22b1a75-eedd-46d0-a848-01110ca19337" />
<img width="522" height="285" alt="Screenshot 2025-12-17 at 2 27 39 PM" src="https://github.com/user-attachments/assets/e66e89ce-d397-42b9-9af1-ff52fd3addef" />
